Add Record encryption to Security Consideration (#306)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

registered. Applications should mitigate against this type of attack by rate
limiting or otherwise restricting the registration flow.

## Protecting the Registration Masking Key

The user enumeration prevention method described in this documents uses a symmetric
encryption key generated by the client on registration that is sent to the server over
an authenticated channel, such as one provided by TLS {{RFC8446}}. In the event that this
channel is compromised, this encryption key could be leaked to an attacker.

One mitigation against this threat is to additionally encrypt the `RegistrationRecord`
sent from client to server at the application layer using public key encryption, e.g.,
with HPKE {{?RFC9180}}.

> 

Reviewer: ```suggestion
```

> However, the threat models and the added complexity for this mechanim are out of scope

Reviewer: ```suggestion
However, the details of this mechanism are out of scope
```

registered. Applications should mitigate against this type of attack by rate
limiting or otherwise restricting the registration flow.

## Protecting the masking key during registration

The user enumeration prevention method uses a symmetric encryption key generated
by the client and sent to the server at registration in the clear. This relatively
sensitive and secret piece of data requires the full registration to be secured

> with integrity, authenticity, and confidentiality, as offered by TLS, for example.

Reviewer 2: I would remove "authenticity" because it is redundant, or if it really needs to be there, at least replace with "authentication"

Author: There's no authentication at the registration phase. The wording could be misleading.

Regarding authenticity, it's with regard to server messages really coming from the server.

Author: Does this answer the question?

With TLS, integrity and authenticity can't be subverted once the session is over
and past. Confidentiality, however, is not ensured if TLS encryption keys are
backed up, shared, or leaked, and this allows decryption after the session (e.g.
due to proxy re-encryption, logging, legal compliance). In transit, the masking
key is subject to logging after TLS termination.

In these cases, the masking key is at risk of being revealed to a passive (even
non-malicious) attacker after registration has happened.

This risk can be avoided by using a Key Encapsulation Mechanism (KEM) like HPKE
{{?I-D.irtf-cfrg-hpke}}. Implementations would use the server's public key received
in the registration response to encrypt the record sent in the last registration

> message from the client to the server.

Author: I'm not sure this really exposes the rationale of the recommandation

registered. Applications should mitigate against this type of attack by rate
limiting or otherwise restricting the registration flow.

## Protecting the Registration Masking Key

The user enumeration prevention method described in this documents uses a symmetric

> encryption key generated by the client on regitration that is sent to the server in
> cleartext. This relatively sensitive and secret piece of data requires the full
> registration to be secured with integrity, authenticity, and confidentiality, as offered
> by TLS, for example.

Reviewer: ```suggestion
encryption key generated by the client on registration that is sent to the server over
an authenticated channel, such as one provided by TLS {{RFC8446}}.
```

> With TLS, integrity and authenticity can't be subverted once the session is over
> and past. Confidentiality, however, is not ensured if TLS encryption keys are
> backed up, shared, or leaked, and this allows decryption after the session (e.g.
> due to proxy re-encryption, logging, legal compliance). In transit, the masking
> key is subject to logging after TLS termination and therefore at risk of being
> revealed to a passive (even non-malicious) attacker after registration has happened.

Reviewer: ```suggestion
In the event that this channel is compromised, this encryption key could be leaked
to an attacker.
```

> Application layer encryption is one possible way to protect this secret in transit.
> 
> Another possible method is to use a Key Encapsulation Mechanism (KEM) like HPKE
> {{?I-D.irtf-cfrg-hpke}} in the application layer. Clients would use the server's
> public key received in the registration response to encrypt the record sent in
> the last registration message from the client to the server.

Reviewer: ```suggestion
One mitigation against this threat is to additionally encrypt the `RegistrationRecord` 
sent from client to server at the application layer using public key encryption, e.g., 
with HPKE {{?RFC9180}}.
```

> It is also possible to avoid using a symmetric encryption key at all by replacing
> the masking mechanism with a non-interactive DHKE to encrypt the envelope in transit
> during registration and the login phases. On registration, the client uses a
> server-provided public key (in the registration response) to encrypt its record to
> that key and send the result to the server. This adds confidentiality to the
> record as discussed above. On login, the server encrypts the envelope to the
> client's public key and sends that along in the credential response. The client
> recovers its private key (just like now) and can decrypt that ciphertext.
> 
> The necessity of these mechanisms is realtive to the thread model the protocol
> is executed in. Their added complexity and are out of scope of this document.

Reviewer: ```suggestion
```

I don't think we can keep this text since it would effectively yield a different protocol. Moreover, this is kind of redundant with the HPKE approach, so I don't think it adds much.

Author: I'm not at all convinced that it yields a different protocol, as this only replaces the envelope encryption, itself an added component to the original OPAQUE protocol. I agree that the added complexity is out of scope for the document though, that's why we said we would mention it here.

cf the discussion in #310

Reviewer: I mean that the protocol mechanics change if this done rather than the existing encryption scheme. I agree that _functionally_ the protocol remains the same, but the implementation would be different. Hence, a different protocol.

Author: Would you accept a shorter, less detailed version of this paragraph stating this rationale? I think we should still mention that there _might_ be a problem for some threat models, and agree the countermeasures burden this spec

Reviewer: Sure, but isn't that already covered by the HPKE suggestion? I don't think any more detail is useful.

Author: Not really, the first addresses TLS compromission, and the second addresses the fact of having a client secret stored on the server. 

I updated the paragraph

Reviewer: I still don't get it. In both cases public key encryption is used. I think the latter case should be removed entirely.

Author: In the first case, the record holding the masking key is encrypted on registration. On login, the current symmetric method is used.

In the second case, the whole symmetric encryption is dropped: the user replaces the symmetric key with a public key on registration (therefore meaning risk number one is thwarted without encryption of the record), and the server encrypts the envelope with that provided key. Here the server does not need to store a secret value.

Both mechanisms use public-key encryption, yes, but only the second one addresses both risks. That's why I'm insisting on its value :)
